### ISSUE TITLE
Support for Razer Ornata V3 [02A1]

### ISSUE THREAD

**Issue by Tankeran** [openrazer/razer-drivers#2053]
### Device Info

* Name: Razer Ornata V3
* Product Number: RZ03-0446
* Model Number: RZ03-04461200-R3L1

### Device Information
```
 Bus 001 Device 004: ID 1532:02a1 Razer USA, Ltd Razer Ornata V3
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x02a1 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Ornata V3
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     179
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
Device Status:     0x0000
  (Bus Powered)
```
```
total 0
drwxr-xr-x 2 root root 280 Apr 24 16:49 .
drwxr-xr-x 4 root root 700 Apr 24 16:49 ..
lrwxrwxrwx 1 root root  10 Apr 24 16:48 usb-Razer_Razer_Ornata_V3-event-if01 -> ../event14
lrwxrwxrwx 1 root root  10 Apr 24 16:48 usb-Razer_Razer_Ornata_V3-event-kbd -> ../event10
lrwxrwxrwx 1 root root  10 Apr 24 16:48 usb-Razer_Razer_Ornata_V3-if01-event-kbd -> ../event12
lrwxrwxrwx 1 root root  10 Apr 24 16:48 usb-Razer_Razer_Ornata_V3-if02-event-mouse -> ../event17
lrwxrwxrwx 1 root root   9 Apr 24 16:48 usb-Razer_Razer_Ornata_V3-if02-mouse -> ../mouse3
lrwxrwxrwx 1 root root  10 Apr 24 16:48 usb-Razer_Razer_Viper_Mini-event-mouse -> ../event11
lrwxrwxrwx 1 root root  10 Apr 24 16:48 usb-Razer_Razer_Viper_Mini-if01-event-kbd -> ../event13
lrwxrwxrwx 1 root root  10 Apr 24 16:48 usb-Razer_Razer_Viper_Mini-if02-event-kbd -> ../event21
lrwxrwxrwx 1 root root   9 Apr 24 16:48 usb-Razer_Razer_Viper_Mini-mouse -> ../mouse2
lrwxrwxrwx 1 root root   9 Apr 24 16:48 usb-Valve_Software_Steam_Controller_123456789ABCDEF-event-mouse -> ../event8
lrwxrwxrwx 1 root root   9 Apr 24 16:48 usb-Valve_Software_Steam_Controller_123456789ABCDEF-if01-event-kbd -> ../event9
lrwxrwxrwx 1 root root   9 Apr 24 16:48 usb-Valve_Software_Steam_Controller_123456789ABCDEF-mouse -> ../mouse4
```

<details>

```
001:004:002:DESCRIPTOR         1682354721.962718
 05 01 09 02 A1 01 09 01 A1 00 05 09 19 01 29 05
 15 00 25 01 75 01 95 05 81 02 75 01 95 03 81 03
 06 00 FF 09 40 75 08 95 02 15 81 25 7F 81 02 05
 01 09 38 15 81 25 7F 75 08 95 01 81 06 09 30 09
 31 16 00 80 26 FF 7F 75 10 95 02 81 06 C0 06 00
 FF 09 02 15 00 25 01 75 08 95 5A B1 01 C0

001:004:001:DESCRIPTOR         1682354721.965704
 05 01 09 06 A1 01 85 01 05 07 05 07 19 E0 29 E7
 15 00 25 01 75 01 95 08 81 02 19 00 29 A0 15 00
 25 01 75 01 95 A0 81 02 05 08 19 01 29 03 15 00
 25 01 75 01 95 03 91 02 95 05 91 01 C0 05 0C 09
 01 A1 01 85 02 19 00 2A 3C 02 15 00 26 3C 02 95
 01 75 10 81 00 75 08 95 13 81 01 C0 05 01 09 80
 A1 01 85 03 19 81 29 83 15 00 25 01 75 01 95 03
 81 02 95 05 81 01 75 08 95 14 81 01 C0 05 01 09
 00 A1 01 85 04 09 03 15 00 26 FF 00 35 00 46 FF
 00 75 08 95 15 81 00 C0 05 01 09 00 A1 01 85 05
 09 03 15 00 26 FF 00 35 00 46 FF 00 75 08 95 15
 81 00 C0

001:004:000:DESCRIPTOR         1682354721.968762
 05 01 09 06 A1 01 05 07 19 E0 29 E7 15 00 25 01
 75 01 95 08 81 02 81 01 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 06 81 00 05 08 19 01 29 03 15 00
 25 01 75 01 95 03 91 02 95 05 91 01 C0

001:003:002:DESCRIPTOR         1682354721.970765
 05 01 09 06 A1 01 05 07 19 E0 29 E7 15 00 25 01
 75 01 95 08 81 02 81 01 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 06 81 00 05 08 19 01 29 03 15 00
 25 01 75 01 95 03 91 02 95 05 91 01 C0

001:003:001:DESCRIPTOR         1682354721.973713
 05 01 09 06 A1 01 85 01 05 07 19 E0 29 E7 15 00
 25 01 75 01 95 08 81 02 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 0E 81 00 C0 05 0C 09 01 A1 01 85
 02 19 00 2A 3C 02 15 00 26 3C 02 95 01 75 10 81
 00 75 08 95 0D 81 01 C0 05 01 09 80 A1 01 85 03
 19 81 29 83 15 00 25 01 75 01 95 03 81 02 95 05
 81 01 75 08 95 0E 81 01 C0 05 01 09 00 A1 01 85
 04 09 03 15 00 26 FF 00 35 00 46 FF 00 75 08 95
 0F 81 00 C0 05 01 09 00 A1 01 85 05 09 03 15 00
 26 FF 00 35 00 46 FF 00 75 08 95 0F 81 00 C0

001:003:000:DESCRIPTOR         1682354721.976687
 05 01 09 02 A1 01 09 01 A1 00 05 09 19 01 29 05
 15 00 25 01 75 01 95 05 81 02 75 01 95 03 81 03
 06 00 FF 09 40 75 08 95 02 15 81 25 7F 81 02 05
 01 09 38 15 81 25 7F 75 08 95 01 81 06 09 30 09
 31 16 00 80 26 FF 7F 75 10 95 02 81 06 C0 06 00
 FF 09 02 15 00 25 01 75 08 95 5A B1 01 C0
```

</details>

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
<Screenshots of all available options as seen on **Razer Synapse** in a Windows Environment>

**Comment by z3ntu** [comment]
This is like the third version of "Ornata V3" with a different USB PID again. Is the device really just called "Ornata V3" or with some other suffix?

Can you please also find the Product Number & Model Number on the device and send them?

**Comment by Tankeran** [comment]
Hello,

Attached you may find product details.

![IMG_1803](https://user-images.githubusercontent.com/131698534/234467394-ed87b842-31fe-4a32-9ce7-1a9f4c279800.JPEG)

**Comment by z3ntu** [comment]
Thanks, I've added the info to the initial comment. Since the model number (RZ03-0446) is the same as #1925 but product number is higher, maybe it's a newer revision of the device? But in any case from openrazer side it's a different USB PID so it needs separate device support.

**Comment by Tankeran** [comment]
Hello z3ntu,

Thanks a lot for the support. Do I need to do anything else?

**Comment by z3ntu** [comment]
USB packet captures from Razer Synapse on Windows would be helpful to have https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol

But otherwise not really I think.

**Comment by Tankeran** [comment]
Hello,
Unfortunately I dont have Windows computer. Only have Steam deck :S

**Comment by Evonder** [comment]
Oh man I was so excited when this was merged and released. I updated on Debian this afternoon but noticed it was still unrecognized, then I noticed I have yet another version of the Ornata V3 X (I know this thread is for the non X ver.). My version is:
```Bus 004 Device 004: ID 1532:02a2 Razer USA, Ltd Razer Ornata V3 X
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x02a2 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Ornata V3 X
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     159
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0010  1x 16 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1

**Comment by z3ntu** [comment]
@Evonder Please open a new issue for that device, thanks.

**Comment by ghost** [comment]
@Evonder and @z3ntu I've created a new issue for the same device (Razer Ornata V3 X) and same product id (02a2). Please have a look at https://github.com/openrazer/openrazer/issues/2059

And sorry for the offtopic.

**Comment by Tankeran** [comment]
any progress on this topic?

**Comment by greasydeal** [comment]
[OrnataV3(0x02A1).tar.gz](https://github.com/openrazer/openrazer/files/12064662/OrnataV3.0x02A1.tar.gz)

Let me know if these will work.

**Comment by z3ntu** [comment]
Seems fine! The device uses transaction id 0x1f, and uses `razer_chroma_extended_matrix_effect_*` methods and looks like `MATRIX_DIMS = [1, 10]` I believe.

Other than that there's not more info in the pcaps. See https://github.com/openrazer/openrazer/blob/master/DEVELOPMENT.md how to add support for this device.

**Comment by greasydeal** [comment]
> Seems fine! The device uses transaction id 0x1f, and uses `razer_chroma_extended_matrix_effect_*` methods and looks like `MATRIX_DIMS = [1, 10]` I believe.
> 
> Other than that there's not more info in the pcaps. See https://github.com/openrazer/openrazer/blob/master/DEVELOPMENT.md how to add support for this device.

I actually forked the project yesterday morning on have my V3 working in my test VM right now. I still need to work out reactive and starlight modes. Should see a PR later this week.

**Comment by xandervill** [comment]
Hello. 
In applications Polychromatic and RazerGenie Ornata V3 it is not defined. Will there be keyboard Ornata V3 support for Linux Ubuntu 22.04? If not, is it possible to use drivers from Ornata V2 and how to do it?
![image](https://github.com/openrazer/openrazer/assets/134795029/e2ce5d1b-5b53-44f8-b914-b7fb13ceedde)

**Comment by z3ntu** [comment]
Someone opened (and closed again) a pull request for this device: #2096 Not sure what the state there is.

**Comment by Tankeran** [comment]
hello friends i just saw there is already shared document by @greasydeal unfortunately I dont know how to try that driver. Could you guide me through?

**Comment by skyline-ch** [comment]
Hello, I have just acquired the Ornata V3 keyboard, I am under Ubuntu 20.04.
The keyboard works well, but it is not recognized by Polychromatic, from what I see, it is due to the version which is not supported by Open Razer (too bad I had checked before buying that I find it in the list)

I can provide you with information from a virtual machine if necessary, or from my Linux, do tests with a little help for complex manipulations.


given on the keyboard:
model no. RZ03-0446
P/NO. RZ03-04460900-R3Q1
in polychromatic id device 1532:02A1
